invoices/sql_migration_test.go

```diff
@@ -0,0 +1,421 @@
+	var timeZones = []*time.Location{
+		time.UTC,
+		time.FixedZone("EST", -5*3600),
+		time.FixedZone("MST", -7*3600),
+		time.FixedZone("PST", -8*3600),
+		time.FixedZone("CEST", 2*3600),
+	}
+
+	// Ensure max is after min
+	if max.Before(min) {
+		min, max = max, min
+	}
+
+	// Calculate the range in nanoseconds
+	duration := max.Sub(min)
+	randDuration := time.Duration(rand.Int63n(duration.Nanoseconds()))
+
+	// Generate the random time
+	randomTime := min.Add(randDuration)
+
+	// Assign a random time zone
+	randomTimeZone := timeZones[rand.Intn(len(timeZones))]
+
+	// Return the time in the random time zone
+	return randomTime.In(randomTimeZone)
+}
+
+// randTime generates a random time between 2009 and 2140.
+func randTime() time.Time {
+	min := time.Date(2009, 1, 3, 0, 0, 0, 0, time.UTC)
+	max := time.Date(2140, 1, 1, 0, 0, 0, 1000, time.UTC)
+
+	return randTimeBetween(min, max)
+}
+
+func randInvoiceTime(invoice *Invoice) time.Time {
+	return randTimeBetween(
+		invoice.CreationDate,
+		invoice.CreationDate.Add(invoice.Terms.Expiry),
+	)
+}
+
+// randHTLCRapid generates a random HTLC for an invoice using rapid to randomize
+// its parameters.
+func randHTLCRapid(t *rapid.T, invoice *Invoice, amt lnwire.MilliSatoshi) (
+	models.CircuitKey, *InvoiceHTLC) {
+
+	htlc := &InvoiceHTLC{
+		Amt:          amt,
+		AcceptHeight: rapid.Uint32Range(1, 999).Draw(t, "AcceptHeight"),
+		AcceptTime:   randInvoiceTime(invoice),
+		Expiry:       rapid.Uint32Range(1, 999).Draw(t, "Expiry"),
+	}
+
+	// Set MPP total amount if MPP feature is enabled in the invoice.
+	if invoice.Terms.Features.HasFeature(lnwire.MPPRequired) {
+		htlc.MppTotalAmt = invoice.Terms.Value
+	}
+
+	// Set the HTLC state and resolve time based on the invoice state.
+	switch invoice.State {
+	case ContractSettled:
+		htlc.State = HtlcStateSettled
+		htlc.ResolveTime = randInvoiceTime(invoice)
+
+	case ContractCanceled:
+		htlc.State = HtlcStateCanceled
+		htlc.ResolveTime = randInvoiceTime(invoice)
+
+	case ContractAccepted:
+		htlc.State = HtlcStateAccepted
+	}
+
+	// Add randomized custom records to the HTLC.
+	htlc.CustomRecords = make(record.CustomSet)
+	numRecords := rapid.IntRange(0, 5).Draw(t, "numRecords")
+	for i := 0; i < numRecords; i++ {
+		key := rapid.Uint64Range(
+			record.CustomTypeStart, 1000+record.CustomTypeStart,
+		).Draw(t, "customRecordKey")
+		value := []byte(randomStringRapid(t, 10))
+		htlc.CustomRecords[key] = value
+	}
+
+	// Generate a unique HTLC ID and assign it to a channel ID.
+	htlcID := atomic.AddUint64(&testHtlcIDSequence, 1)
+	randChanID := lnwire.NewShortChanIDFromInt(htlcID % 5)
+
+	circuitKey := models.CircuitKey{
+		ChanID: randChanID,
+		HtlcID: htlcID,
+	}
+
+	return circuitKey, htlc
+}
+
+// generateInvoiceHTLCsRapid generates all HTLCs for an invoice, including AMP
+// HTLCs if applicable, using rapid for randomization of HTLC count and
+// distribution.
+func generateInvoiceHTLCsRapid(t *rapid.T, invoice *Invoice) {
```

Comment:
I believe in Go it is idiomatic to pass arguments that we mutate by pointer reference instead of value. I agree with your sentiment though. In this case the alternatives would be:

- keep the pointer reference and do not mutate => goes against Go's principles a bit.
- pass by value => it is also not perfect as we don't have copy constructors, so ideally we invent something like that too but it's very error prone.
- pass only the needed parts of the invoice => I might lean towards this, but imho makes the code much less readable overall.

Given this is just test code I don't feel very strong, but enjoy the discussion nonetheless :)